cirq/google/api/v2/program.py

```diff
@@ -93,13 +93,14 @@ def grid_qubit_from_proto_id(proto_id: str) -> 'cirq.GridQubit':
     Raises:
         ValueError: If the string not of the correct format.
     """
-    parts = proto_id.split('_')
-    if len(parts) != 2:
+
+    match = re.match(r'^q?(-?\d+)_(-?\d+)$', proto_id)
```

Comment:
To avoid parsing the regex repeatedly, move to module level and compile once:
```python
GRID_QUBIT_REGEX = re.compile(r'^q?(-?\d+)_(-?\d+)$')

...
def grid_qubit_from_proto_id(proto_id):
    match = GRID_QUBIT_REGEX.match(proto_id)
    ...
```